#ifndef LIGHTSDIALOG_H_
#define LIGHTSDIALOG_H_

#include <array>
#include <QDialog>
#include <ngl/Vec3.h>
#include "Light.h"

//----------------------------------------------------------------------------------------------------------------------
/// @file LightsDialog.h
/// @brief This class contains the user interface required to manage lights in the scene
/// @author Neerav Nagda
/// @version 0.1
/// @date 02/04/17
//----------------------------------------------------------------------------------------------------------------------

namespace Ui {
class LightsDialog;
}

//----------------------------------------------------------------------------------------------------------------------
/// @class LightsDialog
/// @brief This class manages the lights in the scene
//----------------------------------------------------------------------------------------------------------------------
class LightsDialog : public QDialog
{
		Q_OBJECT

	public:
		//----------------------------------------------------------------------------------------------------------------------
		/// @brief Constructor
		/// @param parent The parent object
		//----------------------------------------------------------------------------------------------------------------------
		explicit LightsDialog(QWidget *parent = 0);
		//----------------------------------------------------------------------------------------------------------------------
		/// @brief Destructor
		//----------------------------------------------------------------------------------------------------------------------
		~LightsDialog();
		//----------------------------------------------------------------------------------------------------------------------
		/// @brief Get function for the UI
		/// @return Reference to the UI in this class
		//----------------------------------------------------------------------------------------------------------------------
		const Ui::LightsDialog& getUI() {return *m_ui;}

	signals:
		//----------------------------------------------------------------------------------------------------------------------
		/// @brief Signal when the light is toggled on or off
		/// @param[out] _lightIndex The light to change
		/// @param[out] _status The active status of the light
		//----------------------------------------------------------------------------------------------------------------------
		void lightActive(int _lightIndex, bool _status);
		//----------------------------------------------------------------------------------------------------------------------
		/// @brief Signal when the light position is changed
		/// @param[out] _lightIndex The light to change
		/// @param[out] _position The new position of the light
		//----------------------------------------------------------------------------------------------------------------------
		void positionChanged(int _lightIndex, ngl::Vec3 _position);
		//----------------------------------------------------------------------------------------------------------------------
		/// @brief Signal when the light ambient colour is changed
		/// @param[out] _lightIndex The light to change
		/// @param[out] _ambient The new ambient colour of the light
		//----------------------------------------------------------------------------------------------------------------------
		void ambientChanged(int _lightIndex, ngl::Vec3 _ambient);
		//----------------------------------------------------------------------------------------------------------------------
		/// @brief Signal when the light diffuse colour is changed
		/// @param[out] _lightIndex The light to change
		/// @param[out] _diffuse The new diffuse colour of the light
		//----------------------------------------------------------------------------------------------------------------------
		void diffuseChanged(int _lightIndex, ngl::Vec3 _diffuse);
		//----------------------------------------------------------------------------------------------------------------------
		/// @brief Signal when the light specular colour is changed
		/// @param[out] _lightIndex The light to change
		/// @param[out] _specular The new specular colour of the light
		//----------------------------------------------------------------------------------------------------------------------
		void specularChanged(int _lightIndex, ngl::Vec3 _specular);


	private:
		//----------------------------------------------------------------------------------------------------------------------
		/// @brief The UI of this class
		//----------------------------------------------------------------------------------------------------------------------
		Ui::LightsDialog *m_ui;
		//----------------------------------------------------------------------------------------------------------------------
		/// @brief Set the max number of lights
		/// This must be also defined in the shaders
		//----------------------------------------------------------------------------------------------------------------------
		static constexpr int NumLights = 4;
		//----------------------------------------------------------------------------------------------------------------------
		/// @brief Container of light information
		//----------------------------------------------------------------------------------------------------------------------
		std::array<Light,NumLights> m_lights;
		//----------------------------------------------------------------------------------------------------------------------
		/// @brief Store the currently selected light for faster computation
		//----------------------------------------------------------------------------------------------------------------------
		int m_currentIndex = 0;


	private slots:
		//----------------------------------------------------------------------------------------------------------------------
		/// @brief Change the UI values depending on the light
		/// @param _index The index for the array m_lights
		//----------------------------------------------------------------------------------------------------------------------
		void changeLightInfo(int _index);
		//----------------------------------------------------------------------------------------------------------------------
		/// @brief Change the isActive status
		/// @param _status The new isActive status
		//----------------------------------------------------------------------------------------------------------------------
		void setActiveStatus(bool _status);
		//----------------------------------------------------------------------------------------------------------------------
		/// @brief Set the position
		//----------------------------------------------------------------------------------------------------------------------
		void setPos();
		//----------------------------------------------------------------------------------------------------------------------
		/// @brief Set the ambient
		//----------------------------------------------------------------------------------------------------------------------
		void setAmbient();
		//----------------------------------------------------------------------------------------------------------------------
		/// @brief Set the diffuse
		//----------------------------------------------------------------------------------------------------------------------
		void setDiffuse();
		//----------------------------------------------------------------------------------------------------------------------
		/// @brief Set the specular
		//----------------------------------------------------------------------------------------------------------------------
		void setSpecular();
};

#endif // LIGHTSDIALOG_H_
